File: machine/ts/machine/devices/audio/status_register.ts
import {
	APU_STATUS_BUSY,
	APU_STATUS_CMD_FIFO_EMPTY,
	APU_STATUS_CMD_FIFO_FULL,
} from './contracts';
import type { DeviceScheduler } from '../../scheduler/device';
import type { ApuCommandFifo } from './command_fifo';
import type { ApuServiceClock } from './service_clock';
import type { ApuSlotBank } from './slot_bank';
import type { DeviceStatusLatch } from '../device_status';

export class ApuStatusRegister {
	public constructor(
		private readonly fault: DeviceStatusLatch,
		private readonly slots: ApuSlotBank,
		private readonly commandFifo: ApuCommandFifo,
		private readonly serviceClock: ApuServiceClock,
		private readonly scheduler: DeviceScheduler,
	) {}

	public static readThunk(context: ApuStatusRegister): number {
		const nowCycles = context.scheduler.currentNowCycles();
		context.serviceClock.synchronize(nowCycles);
		const busy = context.slots.activeMask !== 0 || !context.commandFifo.empty;
		return (context.fault.status
			| (busy ? APU_STATUS_BUSY : 0)
			| (context.commandFifo.empty ? APU_STATUS_CMD_FIFO_EMPTY : 0)
			| (context.commandFifo.full ? APU_STATUS_CMD_FIFO_FULL : 0)
			| context.serviceClock.sampleTransferStatusBits()) >>> 0;
	}
}

File: machine/cpp/machine/devices/audio/status_register.cpp
#include "machine/devices/audio/status_register.h"

#include "machine/devices/audio/command_fifo.h"
#include "spec/audio/apu.h"
#include "machine/devices/audio/service_clock.h"
#include "machine/devices/audio/slot_bank.h"
#include "machine/devices/device_status.h"
#include "machine/scheduler/device.h"

namespace bmsx {

ApuStatusRegister::ApuStatusRegister(
	const DeviceStatusLatch& fault,
	const ApuSlotBank& slots,
	const ApuCommandFifo& commandFifo,
	ApuServiceClock& serviceClock,
	DeviceScheduler& scheduler
)
	: m_fault(fault)
	, m_slots(slots)
	, m_commandFifo(commandFifo)
	, m_serviceClock(serviceClock)
	, m_scheduler(scheduler) {}

Value ApuStatusRegister::readThunk(void* context, [[maybe_unused]] u32 addr, MappedBusSignals) {
	auto& reg = *static_cast<ApuStatusRegister*>(context);
	const i64 nowCycles = reg.m_scheduler.currentNowCycles();
	reg.m_serviceClock.synchronize(nowCycles);
	u32 status = reg.m_fault.status | reg.m_serviceClock.sampleTransferStatusBits();
	if (reg.m_slots.activeMask() != 0u || !reg.m_commandFifo.empty()) {
		status |= APU_STATUS_BUSY;
	}
	if (reg.m_commandFifo.empty()) {
		status |= APU_STATUS_CMD_FIFO_EMPTY;
	}
	if (reg.m_commandFifo.full()) {
		status |= APU_STATUS_CMD_FIFO_FULL;
	}
	return valueNumber(static_cast<double>(status));
}

} // namespace bmsx
